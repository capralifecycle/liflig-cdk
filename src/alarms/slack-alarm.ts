import * as constructs from "constructs"
import * as cloudwatchActions from "aws-cdk-lib/aws-cloudwatch-actions"
import * as iam from "aws-cdk-lib/aws-iam"
import * as lambda from "aws-cdk-lib/aws-lambda"
import * as sns from "aws-cdk-lib/aws-sns"
import { Duration } from "aws-cdk-lib"
import * as path from "path"

export interface SlackAlarmProps {
  projectName: string
  envName: string
  slackChannel: string
  slackUrl: string
}

/**
 * SNS Topic that can be used to action alarms, with a Lambda
 * that will send a message to Slack for the alarm.
 */
export class SlackAlarm extends constructs.Construct {
  public readonly alarmTopic: sns.Topic
  public readonly snsAction: cloudwatchActions.SnsAction

  constructor(scope: constructs.Construct, id: string, props: SlackAlarmProps) {
    super(scope, id)

    this.alarmTopic = new sns.Topic(this, "Topic")

    this.snsAction = new cloudwatchActions.SnsAction(this.alarmTopic)

    const slackLambda = new lambda.Function(this, "Function", {
      code: lambda.Code.fromAsset(
        path.join(__dirname, "../../assets/slack-alarm-lambda"),
      ),
      description:
        "Receives CloudWatch Alarms through SNS and sends a formatted version to Slack",
      handler: "index.handler",
      memorySize: 128,
      runtime: lambda.Runtime.PYTHON_3_8,
      timeout: Duration.seconds(6),
      environment: {
        SLACK_URL: props.slackUrl,
        SLACK_CHANNEL: props.slackChannel,
        PROJECT_NAME: props.projectName,
        ENVIRONMENT_NAME: props.envName,
      },
    })

    slackLambda.addPermission("InvokePermission", {
      action: "lambda:InvokeFunction",
      principal: new iam.ServicePrincipal("sns.amazonaws.com"),
      sourceArn: this.alarmTopic.topicArn,
    })

    new sns.Subscription(this, "Subscription", {
      endpoint: slackLambda.functionArn,
      protocol: sns.SubscriptionProtocol.LAMBDA,
      topic: this.alarmTopic,
    })
  }
}
