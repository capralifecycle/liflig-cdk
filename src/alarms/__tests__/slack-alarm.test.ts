import "@aws-cdk/assert/jest"
import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import * as secretsmanager from "aws-cdk-lib/aws-secretsmanager"
import { SlackAlarm } from "../slack-alarm"

test("create slack alarm", () => {
  const app = new App()
  const stack = new Stack(app, "Stack")

  const secret = new secretsmanager.Secret(stack, "TestSecret", {
    secretName: "TestSecret",
  })

  new SlackAlarm(stack, "SlackAlarm", {
    envName: "dev",
    projectName: "my-project",
    slackWebhookUrlSecret: secret,
  })

  expect(stack).toMatchCdkSnapshot()
})
