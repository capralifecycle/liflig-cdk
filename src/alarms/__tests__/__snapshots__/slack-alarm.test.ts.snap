// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create slack alarm 1`] = `
Object {
  "Resources": Object {
    "SlackAlarmFunction7A980434": Object {
      "DependsOn": Array [
        "SlackAlarmFunctionServiceRoleC8F1C44C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "70ce0c12add85e704fccf538920602b149182000218c82db5e7d9d9e60afa3e1.zip",
        },
        "Description": "Receives CloudWatch Alarms through SNS and sends a formatted version to Slack",
        "Environment": Object {
          "Variables": Object {
            "ENVIRONMENT_NAME": "dev",
            "PROJECT_NAME": "my-project",
            "SLACK_CHANNEL": "#my-channel",
            "SLACK_URL": "https://hooks.slack.com/services/ABC/DEF/123",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "SlackAlarmFunctionServiceRoleC8F1C44C",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 6,
      },
      "Type": "AWS::Lambda::Function",
    },
    "SlackAlarmFunctionInvokePermission99C6FD8D": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "SlackAlarmFunction7A980434",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "SlackAlarmTopicA880F65C",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "SlackAlarmFunctionServiceRoleC8F1C44C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SlackAlarmSubscriptionD0659B5B": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "SlackAlarmFunction7A980434",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "SlackAlarmTopicA880F65C",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "SlackAlarmTopicA880F65C": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
